### ISSUE TITLE
Julia Bindings for OpenCV

### ISSUE THREAD

**Issue by archit120** [opencv/opencv#16339]
## Bindings and Tests for the Julia Language

* Author: Archit Rungta / @archit120
* Link: [Raised Issue](https://github.com/opencv/opencv/issues/16233)
* Status: **Draft** 
* Platforms: **All** 
* Complexity: 2-3 Man-Months

## Introduction and Rationale

Julia is a popular dynamic language used for scientific and numerical computing, with performance comparable to traditional statically-typed languages. It has powerful libraries for many different tasks such as FluxMl and Zygote. However, the JuliaImages library lacks a lot of functionality. 

There are two existing OpenCV wrappers: https://github.com/JuliaOpenCV and https://github.com/maxruby/OpenCV.jl. Both of these do not work with current version of Julia and have been abandoned for multiple years now. The only way of using OpenCV from Julia right now is to call the Python bindings for OpenCV using PyCall.jl

This is aimed at creating bindings for Julia which are similar in scope and ease-of-use to OpenCV's Python bindings. Furthermore, test files similar to Python should be created.

## Proposed solution

The solution is to build wrapping for Julia similar to how they are constructed for Python. Ref [1]. Julia has a library CxxWrap.jl which provides the needed functionality of creating Julia wrappers in C++

This wrapper will then call the C++ OpenCV code. All functionality of OpenCV would then be exposed to be called from Julia code.

## Impact on existing code, compatibility

The changes will not affect any other part of OpenCV's functionality. A seperate folder for Julia bindings will contain all the related code. Only a flag for make configuration has to be added which can be off by default.

## Possible alternatives

Julia supports an alternative form of wrapping using Cxx.jl Unlike CxxWrap this requires wrapper code to be written in Julia. Another alternative is to build the Julia bindings on top of Python. This method, however, is not as preferable because of extra dependencies and overheads. 

## References

1. [Description of How Python Bindings Work](https://docs.opencv.org/4.1.0/da/d49/tutorial_py_bindings_basics.html)
2. [Some Work on CxxWrap based binding](https://github.com/clynamen/opencv)

**Comment by archit120** [comment]
I want to work on this for GSoC 2020.

**Comment by nmaxwell** [comment]
You could also consider using the xtensor family [1-5]. The basic idea would be to write an xtensor-based C++ wrapper around opencv, then get python, Julia, and R bindings, almost for free. If/when the xtensor folks support Matlab bindings, you would then get those too.

1. [xtensor vision](https://towardsdatascience.com/the-xtensor-vision-552dd978e9ad)
2. [xtensor](https://github.com/xtensor-stack/xtensor)
3. [xtensor-python](https://github.com/xtensor-stack/xtensor-python)
4. [xtensor-julia](https://github.com/xtensor-stack/xtensor-julia)
5. [xtensor-r](https://github.com/xtensor-stack/xtensor-r)

**Comment by vpisarev** [comment]
thank you; moved here: https://github.com/opencv/opencv/wiki/OE-31.-Julia-Bindings

**Comment by vpisarev** [comment]
let's keep it open for discussion. @archit120, we added your idea to GSoC 2020 ideas list:
https://github.com/opencv/opencv/wiki/GSoC_2020. If you are still interested, please, submit your application via summerofcode.withgoogle.com

**Comment by americast** [comment]
@archit120 May I know how you would like to suggest in getting this task automated? That is, having something similar to [gen2.py](https://github.com/opencv/opencv/blob/master/modules/python/src2/gen2.py). The functions to be wrapped are taken from [hdr_parser.py](https://github.com/opencv/opencv/blob/master/modules/python/src2/hdr_parser.py#L7-L22) while debugging.  
Also, it would be great if we could have some docstrings added to [gen2.py](https://github.com/opencv/opencv/blob/master/modules/python/src2/gen2.py).

**Comment by archit120** [comment]
@americast I don't think we need to modify hdr_parser.py. However, It might be better to convert it to Julia to reduce dependencies. For gen2.py, cv2.cpp and pycompat.hpp, I'll have to re-write the template code to work with CxxWrap.jl. Gen2.py can be kept in python aswell. The package folder and the makefiles would have to be written mostly new.

**Comment by vpisarev** [comment]
@archit120, no, please, do not do it. Do not convert hdr_parser.py to Julia. OpenCV needs Python in any case for many things. As long as you convert hdr_parser.py to Julia, you will have to support this "fork". So, please, leave hdr_parser.py as-is. Consider it as the requirement.

**Comment by archit120** [comment]
@vpisarev hdr_parser.py does not need to be converted at all. I just wondered if it would be better to reduce dependencies. Feedback noted though, I will not change it. 

@americast Could you suggest something for me to do in the meantime? The OpenCV GSoC guide says it's better to have code to demonstrate in your proposal but I'm not sure what would be the most appropriate way to do that considering that the nature of my project is very different from most others in the ideas list.

**Comment by alalek** [comment]
BTW, You may learn also MATLAB bindings generator (based on common hdr_parser.py, check opencv_contrib repo).

> it's better to have code to demonstrate in your proposal

You definitely should show tests and examples on Julia with proposed API.

**Comment by archit120** [comment]
Thanks @alalek for the suggestion. That will definitely be a help. I think I can provide a basic stubbed out implementation and/or maybe write a couple of example of how I envisage the complete API would look and work like. Does that sound good?

**Comment by archit120** [comment]
I created a trivial working example here - https://github.com/archit120/JuliaOpenCVBindingTest/

Right now, I'm first trying to convert between Mat and Julia arrays implicitly like the python binding does between numpy and Mat types

**Comment by cloudhan** [comment]
Can you rename files from cv2 to cv, python binding use `cv2` as its namespace simply due to historical reason: there was a `cv` binding for OpenCV prior to `cv2` binding.

**Comment by archit120** [comment]
@cloudhan I wasn't aware of that. Phase 3 is almost a complete re-write and I think I can do it then.
